domaci_maze.py

```diff
@@ -90,495 +77,523 @@ def has_value(self) -> bool:
+        possible_values_one_node = []
+        for action in ACTIONS:
+            possible_values_one_node.append(values[next_node.get_position(), action])
+        possible_values.append(prob * (next_node.get_reward() + gamma * max(possible_values_one_node)))
+    return sum(possible_values) if possible_values else -100  # wall node
 
 
-def value_iteration(env, gamma, eps, v0=None, maxiter=100):
-    v = v0 if v0 is not None else init_values(env)
-    for k in range(maxiter):
-        nv = async_update_all_values(env, values, gamma)
-        err = max([abs(nv[s] - v[s]) for s in v])
+# updates values for one iteration
+def async_update_all_values(env: MazeEnvironment, values, gamma, q_function):
+    for s in values:
+        if q_function:
+            if not env.is_terminal_pos(s[0]):  # q values are (position,action)
+                values[s] = update_q_value(env, s, values, gamma)
+        else:
+            if not env.is_terminal_pos(s):
+                values[s] = update_v_value(env, s, values, gamma)
+    return copy(values)
+
+
+def value_iteration(env, gamma, eps, maxit=100, q_function=False):
+    values = init_q_values(env) if q_function else init_v_values(env)
+    for iteration in range(maxit):
+        # print(f"\nIteration: {iteration + 1}")
+        # print("Old values: ", values)
+        values_copy = copy(values)
+        new_values = async_update_all_values(env, values, gamma, q_function)
+        # print("New values: ", new_values)
+        err = max([abs(new_values[s] - values_copy[s]) for s in values])
         if err < eps:
-            return nv, k
-        v = nv
-    return v, k
-
+            # print("Final error is: ", err)
+            return new_values, iteration + 1
+        # print("Error is: ", err)
+        values = new_values
+    return values, iteration + 1
+
+
+# should return the smallest number action if there are more actions that have the same value
+# if values of both actions 1 and 2 are -1.0 , it returns 1
+def best_action_min_arg(actions_probs):
+    max_probability = max(prob for _, prob in actions_probs)
+    max_probability_elements = [(action, prob) for action, prob in actions_probs if prob == max_probability]
+
+    min_action = min(action for action, _ in max_probability_elements)
+    min_action_element = [(action, prob) for action, prob in max_probability_elements if action == min_action][0][0]
+
+    return min_action_element
+
+
+def greedy_action(env, current_node, values, gamma, q_function=False):
+    if q_function:
+        action_values = []
+        for action in ACTIONS:
+            for next_node, prob in env.get_action_probabilities(current_node, action):
+                possible_values_one_node = []
+                for next_action in ACTIONS:
+                    possible_values_one_node.append(values[next_node.get_position(), next_action])
+                action_values.append((action, prob * (next_node.get_reward() + gamma * max(possible_values_one_node))))

```

Comment:
Sto opet koristimo belmanovu jednacinu kada q value iteration funkcija radi ovo isto?

Reply:
q value iteration i v value iteration su algoritmi za procenu V i Q vrednosti, i oni idu u krug dok ne konvergiraju. Ovde se za te procenjene Q ili V samo jednom prodje i vraca akcija koja je najbolja za to stanje - optimalna akcija. U sustini formula je ista, vraca se samo druga stvar.